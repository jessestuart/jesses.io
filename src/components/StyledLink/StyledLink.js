import styled from 'styled-components'

const defaultLinkColor = 'rgba(199, 153, 255, 0.7)'
const defaultHoverColor = 'rgba(253, 82, 112, 0.7)'

export const generateUnderlineStyles = ({ linkColor, hoverColor }) => `
  text-decoration: none;
  border-bottom: 2px solid ${linkColor || defaultLinkColor};
  position: relative;

  &::before {
    content: '';
    position: absolute;
    width: 100%;
    height: 2px;
    bottom: -2px;
    left: 0;
    background-color: ${hoverColor || defaultHoverColor};
    visibility: hidden;
    -webkit-transform: scaleX(0);
    transform: scaleX(0);
    -webkit-transition: all 0.3s ease-in-out 0s;
    transition: all 0.3s ease-in-out 0s;
  }
  &:hover:before {
    visibility: visible;
    -webkit-transform: scaleX(1);
    transform: scaleX(1);
  }
  &.no-slide:hover:before {
    visibility: hidden;
  }
  &.no-slide:hover {
    color: #7ef1f9;
  }
`

export default styled.span`
  ${props =>
    generateUnderlineStyles({
      linkColor: props.linkColor,
      hoverColor: props.hoverColor,
    })};
`
