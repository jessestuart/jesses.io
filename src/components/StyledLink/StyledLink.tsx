import { Link } from 'gatsby'
import _ from 'lodash'
import React, { ReactNode } from 'react'
import styled from 'styled-components'
import colors from 'utils/colors'

/**
 * Is a given link "external" -- i.e., should it use Gatby's optimized `<Link>`
 * component or a plain anchor tag?
 */
export const isExternalLinkPredicate: (href: string) => boolean = (
  href: string,
) =>
  _.some([
    _.startsWith(href, 'http'),
    _.startsWith(href, 'mailto'),
    _.startsWith(href, '#'),
    _.endsWith(href, 'pdf'),
  ])

interface Props {
  // ===============
  // Required props:
  // ===============
  children: ReactNode | string
  // ===============
  // Optional props:
  // ===============
  className?: string
  hoverColor?: string
  href?: string
  linkColor?: string
}

/**
 * Fancy-looking component for links or otherwise accented text, with a subtle
 * but delightful hover animation.
 * This can be rendered one of three different ways depending on the input props:
 * 1) "Normal" link to another page on the site. This uses Gatsby's `Link`
 *    component for near-instant loading.
 * 2) "External" links to outside URL's. Uses regular anchor tags.
 * 3) Fraudsters that don't link anywhere, but still look snazzy.
 */
const StyledLink = (props: Props) => {
  const { href, children, ...rest } = props
  // Case #1: no link provided.
  // Just render as a fancypants `<span>`.
  if (!href) {
    return <span {...rest}>{children}</span>
  }

  // Slightly janky hack to prevent Gatsby from complaining
  // about `<Link>`'ing to external resources.
  // TODO: There must be a way to programmatically branch on
  //       this... right?
  const isExternalLink = isExternalLinkPredicate(href)
  // Case #2: link is external.
  if (isExternalLink) {
    return <a {..._.pick(props, ['href', 'className'])}>{children}</a>
  }

  // Case #1: link is internal.
  return (
    <Link to={href} {...rest}>
      {children}
    </Link>
  )
}

export default styled(StyledLink)`
  border-bottom: 2px solid ${props => props.linkColor || colors.defaultLink};
  position: relative;
  text-decoration: none;

  &::before {
    background-color: ${props => props.hoverColor || colors.defaultHover};
    bottom: -2px;
    content: '';
    height: 2px;
    left: 0;
    position: absolute;
    transform: scaleX(0);
    transition: all 0.25s ease-in-out 0s;
    visibility: hidden;
    width: 100%;
  }
  &:hover:before {
    transform: scaleX(1);
    visibility: visible;
  }
`
