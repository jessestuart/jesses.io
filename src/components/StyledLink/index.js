export { default as StyledLink } from './StyledLink'
