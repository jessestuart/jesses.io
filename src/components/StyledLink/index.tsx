import StyledLink from './StyledLink'
export default StyledLink
export { default as StyledLink } from './StyledLink'
