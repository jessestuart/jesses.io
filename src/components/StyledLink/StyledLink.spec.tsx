import React from 'react'
import { StyledLink } from './StyledLink'
import renderer from 'react-test-renderer'

test('Render StyledLink component with default colors.', () => {
  const tree = renderer.create(<StyledLink />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render StyledLink component with custom colors.', () => {
  const tree = renderer
    .create(<StyledLink linkColor="#333333" hoverColor="#555555" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
