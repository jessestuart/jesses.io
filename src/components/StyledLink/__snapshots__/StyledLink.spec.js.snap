// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render StyledLink component with custom colors. 1`] = `
<span
  className="StyledLink-khUpca WzwRW"
/>
`;

exports[`Render StyledLink component with default colors. 1`] = `
<span
  className="StyledLink-khUpca CcbPu"
/>
`;
